solver/src/analytics.rs

```diff
@@ -0,0 +1,123 @@
+use crate::{metrics::SolverMetrics, settlement::Settlement};
+use bigdecimal::{ToPrimitive, Zero};
+use ethcontract::H160;
+use model::order::OrderUid;
+use num::BigRational;
+use shared::conversions::U256Ext;
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    fmt::{Display, Formatter},
+    sync::Arc,
+};
+
+#[derive(Clone)]
+struct SurplusInfo {
+    solver_name: &'static str,
+    absolute: BigRational,
+    ratio: BigRational,
+}
+
+impl Display for SurplusInfo {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "Surplus {{solver: {}, absolute: {:.2e}, ratio: {:.2e} }}",
+            self.solver_name,
+            self.absolute.to_f64().unwrap_or(f64::NAN),
+            self.ratio.to_f64().unwrap_or(f64::NAN)
+        )
+    }
+}
+
+fn get_prices(settlement: &Settlement) -> HashMap<&H160, BigRational> {
+    settlement
+        .clearing_prices()
+        .iter()
+        .map(|(token, price)| (token, price.to_big_rational()))
+        .collect::<HashMap<_, _>>()
+}
+
+/// Record metric with surplus achieved in winning settlement
+/// vs that which was unrealized in other feasible solutions.
+pub fn report_alternative_settlement_surplus(
+    metrics: Arc<dyn SolverMetrics>,
+    winning_settlement: (&'static str, &Settlement),
+    alternative_settlements: Vec<(&'static str, &Settlement)>,
+) {
+    let (winning_solver, submitted) = winning_settlement;
+    let submitted_prices = get_prices(submitted);
+    let submitted_surplus: HashMap<_, _> = submitted
+        .trades()
+        .iter()
+        .map(|trade| {
+            let sell_token_price = &submitted_prices[&trade.order.order_creation.sell_token];
+            let buy_token_price = &submitted_prices[&trade.order.order_creation.buy_token];
+            (
+                trade.order.order_meta_data.uid,
+                SurplusInfo {
+                    solver_name: winning_solver,
+                    absolute: trade
+                        .surplus(sell_token_price, buy_token_price)
+                        .unwrap_or_else(BigRational::zero),
+                    ratio: trade
+                        .surplus_ratio(sell_token_price, buy_token_price)
+                        .unwrap_or_else(BigRational::zero),
+                },
+            )
+        })
+        .collect();
+
+    let best_alternative = best_surplus_by_order(alternative_settlements);
+
+    for (order_id, submitted) in submitted_surplus.iter() {
+        if let Some(alternative) = best_alternative.get(order_id) {
+            metrics.report_order_surplus(
+                winning_solver,
+                alternative.solver_name,
+                (&submitted.ratio - &alternative.ratio)
+                    .to_f64()
+                    .unwrap_or_default(),
+            );
+            if alternative.absolute > submitted.absolute {
```

Comment:
I think you are comparing absolute surpluses without normalizing prices which is dangerous (since just scaling prices will have an effect on this value). For example, when computing the objective value, we scale surplus using external prices. Maybe it needs to be done here as well.

Reply:
Why don't I just compare the ratios here then. That's what they are for.

Reply:
Changed to ratio